### PR TITLE
[wrangler] Widen agent Pages-to-Workers delegation for new projects

### PR DESCRIPTION
**Pull request by MattieTK**
Two related changes that widen the agent Pages-to-Workers delegation so the common "brand-new static project" case lands on Workers instead of falling through to a direct Pages deploy. Both remain agent-only.

## Branch flags no longer disqualify delegation

`wrangler pages deploy --branch <name>` and `wrangler pages project create --production-branch <name>` are now eligible for delegation. A branch flag was implicitly treated as evidence that an established Pages project already existed, so its presence bypassed delegation. But delegation only ever fires for a brand-new project, where a branch flag merely names the production branch – exactly what a Workers static-assets deploy targets. There are no preview-deployment semantics to preserve, so the flag no longer needs to disqualify the command. Genuinely Pages-only flags (`--commit-hash`, `--commit-message`, `--commit-dirty`, `--skip-caching`) still disqualify a deploy.

## Per-project delegation gate

The gate is now per-project rather than per-account. Previously an account that already had any Pages project was never delegated; now a brand-new project is delegated even when the account has other Pages projects, while a deploy to an existing project stays on Pages. This also drops the paginated list-projects API call from the delegation path in favour of the project-existence signal the deploy command already computes.

The experimental rationale telemetry and the opt-out-flag rename from the same research branch are intentionally not included here.

---

- Tests
  - [x] Tests included/updated
  - [ ] Automated tests not possible - manual testing has been completed as follows:
  - [ ] Additional testing not necessary because:
- Public documentation
  - [ ] Cloudflare docs PR(s):
  - [x] Documentation not necessary because: this refines the internal, agent-only Pages-to-Workers delegation heuristic. The opt-out flag and the delegation itself are hidden and not part of user-facing documented behaviour, so there is no public documentation to update.

---

<a href="https://cloudflare.devinenterprise.com/review/cloudflare/workers-sdk/pull/15004" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=3">
    Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- ✅ @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>


</details>

File: packages/wrangler/src/pages/projects.ts [hunk 1]
```diff
@@ -97,6 +97,39 @@ export const listProjects = async ({
...
 };
 
+/**
+ * Whether a Pages project with the given name already exists on the account.
+ *
+ * Returns false for a missing name or a not-found project; any other API error
+ * propagates so the caller can decide how to treat a failed lookup (the
+ * delegation gate skips delegation rather than risk delegating over a project
+ * that may already exist).
+ */
```
Comment (dario-piotrowicz) [inline packages/wrangler/src/pages/projects.ts:117]: nit: personally I love using `@params` and `@returns` as they make the comment more readable and clear

**Review by dario-piotrowicz — APPROVED** [review]
LGTM with a minor nit

**Review by workers-devprod — APPROVED** [review]
Codeowners reviews satisfied

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- ✅ @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>


</details>

**Comment by petebacondarwin** [comment]
We recently landed #14994, a large change to Miniflare's configuration internals that touched ~177 files across the repo. Leaving this PR on its old base was likely to cause conflicts, so we've **rebased it onto the latest `main` and force-pushed** the result.

**Your local copy of this branch is now out of date.** Before you push again, please reset to the new version:

```sh
git fetch origin
git checkout fix/pages-delegation-branch-flags
git reset --hard origin/fix/pages-delegation-branch-flags
```

Because the base moved a long way, it's also worth reinstalling before you carry on — the lockfile changed:

```sh
pnpm install
```

Sorry for the interruption. If the rebase looks wrong, or CI now fails in a way that seems related to the Miniflare config change rather than your own work, comment here and we'll help get it sorted.

File: packages/wrangler/src/pages/projects.ts (hunk 1 above)
Comment (MattieTK) [reply packages/wrangler/src/pages/projects.ts:117]: addressed!

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- ✅ @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>


</details>

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- ✅ @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>


</details>

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- ✅ @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>


</details>

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- ✅ @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>


</details>

File: packages/wrangler/src/pages/deploy.ts [hunk 2]
```diff
@@ -215,14 +215,15 @@ export const pagesDeployCommand = createCommand({
...
+		// using unsupported Pages features, and `--force` are never delegated. The
+		// account is free to already have other Pages projects — only the specific
+		// project being targeted must be new.
 		const delegation = await maybeDelegatePagesToWorkers({
 			command: "deploy",
 			projectPath: process.cwd(),
 			assetsDirectory: directory,
-			accountHasPagesProjects: async () =>
-				(await listProjects({ accountId })).length > 0,
+			projectExists: Boolean(projectName) && isExistingProject,
```
[the remark this answers is not in the document]
Comment (MattieTK) [reply packages/wrangler/src/pages/deploy.ts]: I think this, when isolated to agents only, is acceptable. If the user runs it themselves they will get the project picker.

File: packages/wrangler/src/__tests__/pages/deploy.test.ts [hunk 3]
```diff
@@ -124,6 +129,168 @@ describe("pages deploy", () => {
...
+						messages: [],
+						result: null,
+					},
+					{ status: 404 }
+				);
+			})
+		);
+
+		// This account-only cache write deliberately reproduces the merge that
+		// retains the previous project name while changing its account ID.
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/pages/deploy.test.ts]: I don't really understand this comment.

Why do we need to run `pages project list`?

File: packages/wrangler/src/__tests__/pages/deploy.test.ts [hunk 4]
```diff
@@ -124,6 +129,168 @@ describe("pages deploy", () => {
...
+			runWrangler("pages deploy public")
+		).rejects.toThrowErrorMatchingInlineSnapshot(
+			`[Error: Missing Pages project name. Use --project-name <name> or set the name in your Wrangler configuration file.]`
+		);
+		expect(staleProjectLookupCount).toBe(0);
+		expect(projectListRequestCount).toBe(1);
+		expect(std.out).not.toContain("Delegating to");
+	});
+
+	it("keeps an account-matching cached project on Pages when it is missing remotely", async ({
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/pages/deploy.test.ts]: The idea of this case is that the cache suggests that this project has been previously deployed to Pages, even though not to the current account?

File: packages/wrangler/src/__tests__/pages/deploy.test.ts [hunk 5]
```diff
@@ -124,6 +129,168 @@ describe("pages deploy", () => {
...
+
+		await expect(
+			runWrangler("pages deploy public")
+		).rejects.toThrowErrorMatchingInlineSnapshot(
+			`[Error: Missing Pages project name. Use --project-name <name> or set the name in your Wrangler configuration file.]`
+		);
+		expect(std.out).not.toContain("Delegating to");
+	});
+
+	it("does not delegate using a stale cached project name after an agent switches accounts", async ({
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/pages/deploy.test.ts]: Doesn't this test case cover the scenario where there is a cached project but the user has specified no project name?

File: packages/wrangler/src/__tests__/pages/deploy.test.ts [hunk 6]
```diff
@@ -1967,6 +2134,137 @@ describe("pages deploy", () => {
 		expect(std.err).toMatchInlineSnapshot(`""`);
 	});
 
+	it("preserves preview semantics for an interactive agent creating a new project with --branch", async ({
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/pages/deploy.test.ts:2138]: Can you explain this test case?
What are the preview semantics we are testing here?
Are you saying that if we set the `--branch` option we should not delegate?
Is this because the Workers Previews feature is not fully live yet?

File: packages/wrangler/src/__tests__/pages/deploy.test.ts [hunk 7]
```diff
@@ -6650,3 +6948,50 @@ function mockGetProjectHandler(
...
+		]);
+	});
+
+	it("ignores boolean flags left at false", ({ expect }) => {
+		const args = { commitDirty: false, skipCaching: false } as DeployArgs;
+
+		expect(getUnsupportedDeployDelegateArgs(args)).toEqual([]);
+	});
+
+	it("reports only the genuinely unsupported flags when a branch is combined with commit metadata", ({
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/pages/deploy.test.ts]: Can you explain this test-case?
What do you mean by "genuine"?

**Review by petebacondarwin — COMMENTED** [review]
I feel like we are missing tests that show we will still delegate a new pages project when there are other pages projects in the account?

File: packages/wrangler/src/__tests__/pages/deploy.test.ts (hunk 5 above)
Comment (MattieTK) [reply packages/wrangler/src/__tests__/pages/deploy.test.ts]: This is the account mismatch test: will rename to that. If user has a pages project on account A, then switches to account B, we should not delegate the project.

File: packages/wrangler/src/__tests__/pages/deploy.test.ts (hunk 4 above)
Comment (MattieTK) [reply packages/wrangler/src/__tests__/pages/deploy.test.ts]: Clanker says we only save the resulting account_id with saveToConfigCache so we may end up in a situation with a project that doesn't exist in the account if you run a command that rewrites the cache before running your pages deployment on account B for a pages project deployed on account A. Will rename to make clearer

File: packages/wrangler/src/__tests__/pages/deploy.test.ts (hunk 3 above)
Comment (MattieTK) [reply packages/wrangler/src/__tests__/pages/deploy.test.ts]: This is the edge case above, if the user runs a command like pages project list in a directory on account B, after being logged into the project-owning account A, then they will rewrite the account_id only in the pages state file, making it appear that account B has a project with that name.

File: packages/wrangler/src/__tests__/pages/deploy.test.ts (hunk 7 above)
Comment (MattieTK) [reply packages/wrangler/src/__tests__/pages/deploy.test.ts]: This isn't very important on further reflection, we have tests above that test branch and commit options independently, so I think it's safe to remove.

The case was supposed to be for clarifying that we are not able to do anything with these deploy arguments in the delegation because Workers has no equivalent.

Sorry for the 'genuine', which has known unknown origins.

File: packages/wrangler/src/__tests__/pages/deploy.test.ts (hunk 6 above)
Comment (MattieTK) [reply packages/wrangler/src/__tests__/pages/deploy.test.ts:2138]: Yes to all this. When Workers Previews goes live we could be in a position to deploy these also, but I also think it likely suggests the project already exists. Workers Previews I don't think has a feature that allows you to deploy _only_ a preview, which is this suggestion (eg. leave my main worker empty). @yomna-shousha, I don't think this is essential but tagging you just in case 😄
